packages/@react-spectrum/s2/src/CardView.tsx

```diff
@@ -502,28 +510,55 @@ const cardViewStyles = style({
   outlineOffset: -2
 }, getAllowedOverrides({height: true}));
 
-export function CardView<T extends object>(props: CardViewProps<T>) {
-  let {children, layout: layoutName = 'grid', size = 'M', density = 'regular', variant = 'primary', selectionStyle = 'checkbox', UNSAFE_className = '', UNSAFE_style, styles, ...otherProps} = props;
-  let options = layoutOptions[size][density];
+function CardView<T extends object>(props: CardViewProps<T>, ref: DOMRef<HTMLDivElement>) {
+  let {children, layout: layoutName = 'grid', size: sizeProp = 'M', density = 'regular', variant = 'primary', selectionStyle = 'checkbox', UNSAFE_className = '', UNSAFE_style, styles, ...otherProps} = props;
+  let domRef = useDOMRef(ref);
   let layout = useMemo(() => {
-    variant; // needed to invalidate useMemo
-    return layoutName === 'waterfall' ? new WaterfallLayout(options) : new FlexibleGridLayout(options);
-  }, [options, variant, layoutName]);
+    return layoutName === 'waterfall' ? new WaterfallLayout() : new FlexibleGridLayout();
+  }, [layoutName]);
+  
+  // This calculates the maximum t-shirt size where at least two columns fit in the available width.
+  let [maxSizeIndex, setMaxSizeIndex] = useState(SIZES.length - 1);
+  let updateSize = useEffectEvent(() => {
+    let w = domRef.current?.clientWidth ?? 0;
+    let i = SIZES.length - 1;
+    while (i > 0) {
+      let opts = layoutOptions[SIZES[i]][density];
+      if (w >= opts.minItemSize.width * 2 + opts.minSpace.width * 3) {
+        break;
+      }
+      i--;
+    }
+    setMaxSizeIndex(i);
+  });
+
+  useResizeObserver({
+    ref: domRef,
+    box: 'border-box',
+    onResize: updateSize
+  });
+
+  useLayoutEffect(() => {
```

Comment:
https://github.com/adobe/react-spectrum/issues/6091 ?
(followup? trying to decide if we want to do this)